packages/core/src/tools/mcp-list-resources-tool.ts

```diff
@@ -0,0 +1,132 @@
+/**
+ * @license
+ * Copyright 2025 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import type { MessageBus } from '../confirmation-bus/message-bus.js';
+import type { Config } from '../config/config.js';
+import type { ToolResult } from './tools.js';
+import { BaseDeclarativeTool, BaseToolInvocation, Kind } from './tools.js';
+import { LIST_RESOURCES_TOOL_NAME } from './tool-names.js';
+import type { DiscoveredMCPResource } from '../resources/resource-registry.js';
+
+export interface ListResourcesToolParams {
+  /**
+   * Optional: limit results to a single MCP server.
+   */
+  server_name?: string;
+}
+
+class ListResourcesToolInvocation extends BaseToolInvocation<
+  ListResourcesToolParams,
+  ToolResult
+> {
+  constructor(
+    private readonly config: Config,
+    params: ListResourcesToolParams,
+    messageBus?: MessageBus,
+    toolName?: string,
+    toolDisplayName?: string,
+  ) {
+    super(params, messageBus, toolName, toolDisplayName);
+  }
+
+  getDescription(): string {
+    if (this.params.server_name) {
+      return `List resources for ${this.params.server_name}`;
+    }
+    return 'List all discovered MCP resources';
+  }
+
+  private buildSummary(resources: DiscoveredMCPResource[]): string {
+    if (resources.length === 0) {
+      if (this.params.server_name) {
+        return `No MCP resources are currently available for server "${this.params.server_name}".`;
+      }
+      return 'No MCP resources are currently available.';
+    }
+
+    const lines = resources.map((resource) => {
```

Comment:
Doesn't have to happen here, but worth considering what happens if there are e.g., thousands of resources in a follow-up. Would you mind adding a TODO?